backend/service/service_asset.go

```diff
@@ -0,0 +1,139 @@
+package service
+
+import (
+	"github.com/irisnet/explorer/backend/lcd"
+	"github.com/irisnet/explorer/backend/logger"
+	"github.com/irisnet/explorer/backend/model"
+	"github.com/irisnet/explorer/backend/orm/document"
+	"github.com/irisnet/explorer/backend/types"
+	"github.com/irisnet/explorer/backend/utils"
+	"gopkg.in/mgo.v2/bson"
+	"gopkg.in/mgo.v2/txn"
+	"sync"
+)
+
+type AssetService struct {
+	BaseService
+}
+
+func (service *AssetService) GetModule() Module {
+	return Validator
+}
+
+func (service *AssetService) UpdateAssetTokens(vs []document.Asset) error {
+	var vMap = make(map[string]document.Asset)
+	for _, v := range vs {
+		vMap[v.Symbol] = v
+	}
+
+	dstAssetTokens := buildAssetTokens()
+	var txs []txn.Op
+	for _, v := range dstAssetTokens {
+		if it, ok := vMap[v.Symbol]; ok {
+			if isDiffAssetToken(it, v) {
+				v.ID = it.ID
+				txs = append(txs, txn.Op{
+					C:  document.CollectionNmAsset,
+					Id: it.ID,
+					Update: bson.M{
+						"$set": v,
+					},
+				})
+			}
+			delete(vMap, v.Symbol)
+		} else {
+			v.ID = bson.NewObjectId()
+			txs = append(txs, txn.Op{
+				C:      document.CollectionNmAsset,
+				Id:     bson.NewObjectId(),
+				Insert: v,
+			})
+		}
+	}
+
+	if len(vMap) > 0 {
+		for symbol := range vMap {
+			v := vMap[symbol]
+			txs = append(txs, txn.Op{
+				C:      document.CollectionNmAsset,
+				Id:     v.ID,
+				Remove: true,
+			})
+		}
+	}
+	return document.Asset{}.Batch(txs)
+}
+
+func buildAssetTokens() []document.Asset {
```

Comment:
there is no need use goroutine to build asset tokens

Reply:
remove gorutine